### PR TITLE
feat(website): setup sidebar labels and position

### LINKED ISSUE #921: website: build customized sidebar for the documents
# Prerequisites

Please answer the following question for yourself before submitting a feature request.

- [ ] I checked to make sure that this request has not already been filed

# Motivation

Presently, we employ *Docusaurus*'s auto-generated sidebar, which doesn't appear optimal for solidity-docs and typedoc. We have the option to construct a custom sidebar that better communicates the content of each file to enhance the user experience.

# Details
image

➡️ The **solidity-docs** could be changed to **Solidity Docs**, while **index** should be changed to **Typedoc**.

image
image

➡️ Names here should also be modified to readable titles.

Make sure that you have reviewed [MACI's Contributor Guidelines](https://github.com/privacy-scaling-explorations/maci/blob/dev/CONTRIBUTING.md)

### PR DESCRIPTION
**Pull request by kittybest**
# Description

Setup sidebar position and label by adding scripts.

- create a `typedoc.base.json` file in the root folder, for the `typedoc.json` file in each package to extend from.
- adding `typedoc.json` to the `core`, `cli`, `crypto`, `domainobjs` directories, and adding scripts to gen typedoc in their `package.json` ➡️ to separate the classes and interfaces, since before they were put together and it's not easy for users to read. Also now we could have the `README.md` file as the **index page** for each package.
- rewrite the whole `setupTypedoc.ts`, since the typedoc-generating output is totally different now.
- add sidebar config to the auto-generated files, to make the sidebar looks more readable.
- update `core/READMD.md` ➡️ although it's not directly related to this PR, the links in the file is no longer available since we refactor the whole file structure a lot, which causing the tests of **building website** fail.
 

## Related issue(s)

close #921 

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](https://maci.pse.dev/docs/contributing) and [code of conduct](https://maci.pse.dev/docs/contributing/code-of-conduct).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847).
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

**Comment by kittybest** [comment]
Remaining problem is the index page for **solidity docs** and **typedoc**, they're empty now, but I think there should be some content indicating what these following files are.

**Comment by ctrlc03** [comment]
> Remaining problem is the index page for **solidity docs** and **typedoc**, they're empty now, but I think there should be some content indicating what these following files are.

Are you going to tackle that in a separate PR?

File: website/src/scripts/setupTypedoc.ts [hunk 1]
```diff
@@ -1,45 +1,67 @@
...
-  });
+function updateMdFiles(file: string, sidebarInfo: string) {
+  const content = fs.readFileSync(file).toString().split("\n");
+  if (content.length > 3) {
+    content.shift();
+    content.shift();
+    content.shift();
+  }
+  const writtenContent = content.join("\n").replaceAll("README.md", "index.md");
+  fs.writeFileSync(file, `${sidebarInfo}\n${writtenContent}`);
```
Comment (0xmad) [inline website/src/scripts/setupTypedoc.ts:23]:
```suggestion
  const content = fs.readFileSync(file).toString().split("\n");

  if (content.length > 3) {
    content.shift();
    content.shift();
    content.shift();
  }

  const writtenContent = content.join("\n").replaceAll("README.md", "index.md");
  fs.writeFileSync(file, `${sidebarInfo}\n${writtenContent}`);
```

File: website/src/scripts/setupTypedoc.ts [hunk 2]
```diff
@@ -1,45 +1,67 @@
...
 
-// Remove the README.md file if exists
-const readmeFile = path.join(TYPEDOC_DIR, "README.md");
-if (fs.existsSync(readmeFile)) {
-  fs.unlinkSync(readmeFile);
-}
+// read all dir in typedoc/ -> rename README.md as index.md -> remove upper navigations
+const directories = fs.readdirSync(TYPEDOC_DIR);
+directories.forEach((dir) => {
+  const dirname = path.join(TYPEDOC_DIR, dir);
```
Comment (0xmad) [inline website/src/scripts/setupTypedoc.ts:29]: Let's use path resolve instead of join.

File: website/src/scripts/utils.ts [hunk 3]
```diff
@@ -28,3 +28,55 @@ export function copyDirectory(source: string, target: string): void {
...
+ * @param str - the string being transformed
+ */
+export function fitFormat(str: string): string {
+  return str[0].toUpperCase() + str.slice(1, str.length).toLowerCase();
+}
+
+/**
+ * Sidebar Parameters
+ */
+interface sidebarProps {
```
Comment (0xmad) [inline website/src/scripts/utils.ts:44]:
```suggestion
interface SidebarProps {
```

File: website/src/scripts/utils.ts [hunk 4]
```diff
@@ -28,3 +28,55 @@ export function copyDirectory(source: string, target: string): void {
...
+  }
+  if (label) {
+    ret = `${ret}sidebar_label: ${label}\n`;
+  }
+  if (position) {
+    ret = `${ret}sidebar_position: ${position}\n`;
+  }
+  ret = `${ret}---\n`;
+
+  return ret;
```
Comment (0xmad) [inline website/src/scripts/utils.ts:72]: Let's add new lines between conditions.

File: website/src/scripts/utils.ts [hunk 5]
```diff
@@ -28,3 +28,55 @@ export function copyDirectory(source: string, target: string): void {
...
   });
 }
+
+/**
+ * A function that forces the first letter to be capital,
+ * while the rest of the string must be lower case.
+ * @param str - the string being transformed
+ */
+export function fitFormat(str: string): string {
+  return str[0].toUpperCase() + str.slice(1, str.length).toLowerCase();
```
Comment (0xmad) [inline website/src/scripts/utils.ts:38]:
```suggestion
  return `${str[0].toUpperCase()}${str.slice(1, str.length).toLowerCase()}`;
```

File: website/src/scripts/setupTypedoc.ts [hunk 6]
```diff
@@ -1,45 +1,67 @@
...
+    if (fs.existsSync(readmeFile)) {
+      updateMdFiles(readmeFile, generateSidebarString({ title: label, label }));
+      fs.renameSync(readmeFile, path.join(dirname, "index.md"));
+    }
+
+    // remove the first two lines of navigator
+    const modulesFile = path.join(dirname, "modules.md");
+    if (fs.existsSync(modulesFile)) {
+      updateMdFiles(modulesFile, generateSidebarString({ title: `${label} Module`, label: "module", position: 1 }));
+    }
```
Comment (0xmad) [inline website/src/scripts/setupTypedoc.ts:44]:
```suggestion
  if (fs.statSync(dirname).isDirectory()) {
    const readmeFile = path.resolve(dirname, "README.md");

    if (fs.existsSync(readmeFile)) {
      updateMdFiles(readmeFile, generateSidebarString({ title: label, label }));
      fs.renameSync(readmeFile, path.join(dirname, "index.md"));
    }

    // remove the first two lines of navigator
    const modulesFile = path.resolve(dirname, "modules.md");

    if (fs.existsSync(modulesFile)) {
      updateMdFiles(modulesFile, generateSidebarString({ title: `${label} Module`, label: "module", position: 1 }));
    }
```

File: website/src/scripts/setupTypedoc.ts [hunk 7]
```diff
@@ -1,45 +1,67 @@
...
+        const innerFiles = fs.readdirSync(innerDirname);
+        innerFiles.forEach((innerFile) => {
+          const innerLabel = innerFile.split(".")[0];
+          updateMdFiles(
+            path.join(innerDirname, innerFile),
+            generateSidebarString({ title: innerLabel, label: innerLabel }),
+          );
+        });
+      }
+    });
```
Comment (0xmad) [inline website/src/scripts/setupTypedoc.ts:59]:
```suggestion
    const innerDirs = fs.readdirSync(dirname);

    innerDirs.forEach((innerDir) => {
      const innerDirname = path.join(dirname, innerDir);

      if (fs.statSync(innerDirname).isDirectory()) {
        const innerFiles = fs.readdirSync(innerDirname);
        innerFiles.forEach((innerFile) => {
          const innerLabel = innerFile.split(".")[0];
          updateMdFiles(
            path.join(innerDirname, innerFile),
            generateSidebarString({ title: innerLabel, label: innerLabel }),
          );
        });
      }
    });
```

**Review by 0xmad — CHANGES_REQUESTED** [review]
@kittybest thanks, just few comments about code conventions.

**Comment by kittybest** [comment]
> > Remaining problem is the index page for **solidity docs** and **typedoc**, they're empty now, but I think there should be some content indicating what these following files are.
> 
> Are you going to tackle that in a separate PR?

Yes I think so, since there's also another separate issue #922 . This PR could just for #921.

File: website/src/scripts/setupTypedoc.ts (hunk 2 above)
Comment (kittybest) [reply website/src/scripts/setupTypedoc.ts:29]: For all of them, or just this?

Comment (0xmad) [reply website/src/scripts/setupTypedoc.ts]: Right, for all of them.

**Review by 0xmad — APPROVED** [review]
@kittybest thanks!

**Pull request merged**